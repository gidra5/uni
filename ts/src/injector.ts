import { FileMap } from "codespan-napi";
import type { Precedence } from "./ast";
import type { Position } from "./position";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  ASTNodeNextId = "ASTNodeNextId",
  ASTNodePrecedenceMap = "ASTNodePrecedenceMap",
  ASTNodePositionMap = "ASTNodePositionMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.ASTNodeNextId]: number;
  [Injectable.ASTNodePrecedenceMap]: Map<string, Precedence>;
  [Injectable.ASTNodePositionMap]: Map<string, Position>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.ASTNodeNextId, 0);
register(Injectable.ASTNodePrecedenceMap, new Map());
register(Injectable.ASTNodePositionMap, new Map());

export { register, inject, Injectable };
