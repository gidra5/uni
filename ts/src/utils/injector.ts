import { FileMap } from "codespan-napi";
import type { Precedence } from "../ast";
import type { Position } from "./position";
import { PhysicalType, Type } from "../analysis/types/utils";

enum Injectable {
  FileMap = "FileMap",
  RootDir = "RootDir",
  NextId = "NextId",
  PrecedenceMap = "PrecedenceMap",
  PositionMap = "PositionMap",
  TypeMap = "TypeMap",
  PhysicalTypeMap = "PhysicalTypeMap",
  FreeVariablesMap = "FreeVariablesMap",
  BoundVariablesMap = "BoundVariablesMap",
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
  [Injectable.NextId]: number;
  [Injectable.PrecedenceMap]: Map<number, Precedence>;
  [Injectable.PositionMap]: Map<number, Position>;
  [Injectable.TypeMap]: Map<number, Type>;
  [Injectable.PhysicalTypeMap]: Map<number, PhysicalType>;
  [Injectable.FreeVariablesMap]: Map<number, string[]>;
  [Injectable.BoundVariablesMap]: Map<number, string[]>;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(name: T, value: InjectableType[T]) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());
register(Injectable.NextId, 0);
register(Injectable.PrecedenceMap, new Map());
register(Injectable.PositionMap, new Map());
register(Injectable.TypeMap, new Map());

export { register, inject, Injectable };
