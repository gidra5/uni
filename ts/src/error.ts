import { Diagnostic, LabelInfo, primaryDiagnosticLabel, secondaryDiagnosticLabel } from "codespan-napi";
import { assert } from "./utils/index.js";
import { Position } from "./position.js";
import { inject, Injectable } from "./injector.js";
import { getExprPrecedence, NodeType } from "./ast.js";

export enum ErrorType {
  UNKNOWN,
  END_OF_SOURCE,
  UNTERMINATED_STRING,
  INVALID_BINARY_LITERAL,
  INVALID_OCTAL_LITERAL,
  INVALID_HEX_LITERAL,

  MISSING_TOKEN,
  INVALID_PATTERN,
  INVALID_TUPLE_PATTERN,
  INVALID_APPLICATION_EXPRESSION,
  INVALID_TOKEN_EXPRESSION,
  UNDECLARED_NAME_ASSIGNMENT,
  INVALID_PLACEHOLDER_EXPRESSION,
  EVALUATION_ERROR,
  UNDECLARED_NAME,
  INVALID_OBJECT_PATTERN,
  IMPORT_FAILED,
  IMPORT_RESOLVE_FAILED,
  INVALID_INCREMENT_ASSIGN,
  DUPLICATE_DEFAULT_EXPORT,
  INVALID_DEFAULT_PATTERN,
  INVALID_PIN_PATTERN,
  IMMUTABLE_VARIABLE_ASSIGNMENT,
  MISPLACED_OPERATOR,
  MISSING_OPERAND,
  UNCLOSED_BLOCK_COMMENT,
}

type Options = {
  cause?: unknown;
  fileId?: number;
  data?: Record<string, any>;
  notes?: string[];
  labels?: ErrorLabel[];
};

type ErrorLabel = LabelInfo & {
  kind: "primary" | "secondary";
  fileId?: number;
};

export class SystemError extends Error {
  data: Record<string, any>;
  fileId?: number;
  readonly type: ErrorType;
  private labels: ErrorLabel[];
  private notes: string[];
  private constructor(type: ErrorType, msg: string, options: Options = {}) {
    super(msg, { cause: options.cause });
    this.data = options.data ?? {};
    this.fileId = options.fileId;
    this.type = type;
    this.notes = options.notes ?? [];
    this.labels = options.labels ?? [];
  }

  toObject(): any {
    return { ...this, type: ErrorType[this.type], message: this.message };
  }

  withFileId(fileId: number): SystemError {
    this.fileId = fileId;
    return this;
  }

  withCause(cause: unknown): SystemError {
    this.cause = cause;

    if (this.cause instanceof SystemError) {
      const causeMsg = this.cause.message;
      const causeErrorType = ErrorType[this.cause.type];
      this.notes.push(`Caused by: [${causeErrorType}] ${causeMsg}`);
    }

    return this;
  }

  withLabel(kind: "secondary" | "primary", message: string, position: Position, fileId?: number) {
    if (fileId === undefined) this.labels.push({ ...position, message, kind });
    else this.labels.push({ ...position, message, kind, fileId });

    return this;
  }

  withPrimaryLabel(message: string, position: Position, fileId?: number) {
    return this.withLabel("primary", message, position, fileId);
  }

  withSecondaryLabel(message: string, position: Position, fileId?: number) {
    return this.withLabel("secondary", message, position, fileId);
  }

  withNote(message: string) {
    this.notes.push(message);
    return this;
  }

  withNotes(notes: string[]) {
    this.notes = notes;
    return this;
  }

  print(): SystemError {
    const fileMap = inject(Injectable.FileMap);
    const diag = this.diagnostic();
    diag.emitStd(fileMap);
    return this;
  }

  diagnostic(): Diagnostic {
    assert(this.fileId !== undefined, "fileId is not set for SystemError");
    const id = this.fileId;
    const diag = Diagnostic.error();
    const labels = this.labels.map(({ kind, fileId, ...label }) =>
      kind === "primary" ? primaryDiagnosticLabel(fileId ?? id, label) : secondaryDiagnosticLabel(fileId ?? id, label)
    );

    diag.withMessage(this.message);
    diag.withCode(ErrorType[this.type]);
    diag.withLabels(labels);
    diag.withNotes(this.notes);

    return diag;
  }

  static unknown(): SystemError {
    return new SystemError(ErrorType.UNKNOWN, "Unknown error");
  }

  static endOfSource(pos: Position): SystemError {
    return new SystemError(ErrorType.END_OF_SOURCE, "Unexpected end of source").withPrimaryLabel("here", pos);
  }

  static unterminatedString(pos: Position): SystemError {
    return new SystemError(ErrorType.UNTERMINATED_STRING, "Unterminated string literal")
      .withPrimaryLabel("expected closing double quote", pos)
      .withNote('Strings must be enclosed in double quotes (")')
      .withNote("Use \\ to escape special characters");
  }

  static invalidBinaryLiteral(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_BINARY_LITERAL, "Invalid binary literal")
      .withPrimaryLabel("expected digits 0 or 1", {
        start: pos.start + 2,
        end: pos.start + 3,
      })
      .withNote(
        "Valid binary literals start with 0b and digits 0 or 1 (binary digits), which may be followed by more binary digits, optionally separated by underscores"
      );
  }

  static unclosedBlockComment(pos: Position): SystemError {
    return new SystemError(ErrorType.UNCLOSED_BLOCK_COMMENT, "Block comment is not closed")
      .withPrimaryLabel('expected block comment to be closed by "*/"', pos)
      .withSecondaryLabel('insert "*/" here', { start: pos.end - 1, end: pos.end });
  }

  static invalidOctalLiteral(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_OCTAL_LITERAL, "Invalid octal literal")
      .withPrimaryLabel("expected a digit between 0 and 7", {
        start: pos.start + 2,
        end: pos.start + 3,
      })
      .withNote(
        "Valid octal literals start with 0o and a digit between 0 and 7 (octal digits), which may be followed by more octal digits, optionally separated by underscores"
      );
  }

  static invalidHexLiteral(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_HEX_LITERAL, "Invalid hex literal")
      .withPrimaryLabel("expected a digit or a letter between a and f (case insensitive)", {
        start: pos.start + 2,
        end: pos.start + 3,
      })
      .withNote(
        "Valid hex literals start with 0x and a digit or a case insensitive letter between a and f (hex digits), which may be followed by more hex digits, optionally separated by underscores"
      );
  }

  static missingToken(pos: Position, ...tokens: string[]): SystemError {
    const list = tokens.map((token) => `"${token}"`).join(" or ");
    return new SystemError(ErrorType.MISSING_TOKEN, `Missing token, expected one of: ${list}`).withPrimaryLabel(
      "somewhere here",
      pos
    );
  }

  static invalidPattern(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_PATTERN, "invalid pattern").withPrimaryLabel("here", pos);
  }

  static invalidPlaceholderExpression(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_PLACEHOLDER_EXPRESSION, "placeholder can't be evaluated").withPrimaryLabel(
      "here",
      pos
    );
  }

  static undeclaredNameAssignment(name: string, pos: Position, closestName?: string): SystemError {
    const error = new SystemError(ErrorType.UNDECLARED_NAME_ASSIGNMENT, `can't assign to undeclared variable`)
      .withPrimaryLabel(`variable "${name}" is not declared in scope`, pos)
      .withNote("Variable must be declared before it can be assigned to.")
      .withNote("Use := operator to declare a new variable, = assigns to already declared variables only.");

    return closestName
      ? error.withNote(`Did you mean "${closestName}"?`)
      : error.withNote(`Check if you have a typo in the variable name, if "${name}" is intended to be declared.`);
  }

  static invalidTuplePattern(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_TUPLE_PATTERN, "tuple pattern on non-tuple").withPrimaryLabel("here", pos);
  }

  static evaluationError(msg: string, notes: string[], pos: Position): SystemError {
    return new SystemError(ErrorType.EVALUATION_ERROR, msg)
      .withPrimaryLabel("can't evaluate this expression", pos)
      .withNotes(notes);
  }

  static invalidArgumentType(
    name: string,
    signature: { args: [label: string, type: string][]; returns: string },
    pos: Position
  ) {
    return (argIndex: number) => {
      const argSignature = signature.args[argIndex];
      const msg = `${name} ${argSignature[0]} expected ${argSignature[1]}`;
      const argNote = `${name} expects ${argSignature[1]} ${argSignature[0]} as the ${argIndex + 1} argument`;
      const signatureStringifiedArgs = signature.args.map(([label, type]) => `${label}: ${type}`).join(", ");
      const signatureNote = `${name} signature is: (${signatureStringifiedArgs}) => ${signature.returns}`;
      return SystemError.evaluationError(msg, [argNote, signatureNote], pos);
    };
  }

  static invalidIndexTarget(pos: Position): SystemError {
    return SystemError.evaluationError("index operator expects a list or record value on the left side", [], pos);
  }

  static invalidIndex(pos: Position): SystemError {
    return SystemError.evaluationError("index operator expects an integer, string or symbol", [], pos);
  }

  static invalidUseOfSpread(pos: Position): SystemError {
    return SystemError.evaluationError("spread operator can only be used during tuple construction", [], pos);
  }

  static invalidSendChannel(pos: Position): SystemError {
    return SystemError.evaluationError("send operator expects a channel on the left side", [], pos);
  }

  static invalidReceiveChannel(pos: Position): SystemError {
    return SystemError.evaluationError("receive operator expects a channel on the right side", [], pos);
  }

  static channelClosed(pos: Position): SystemError {
    return SystemError.evaluationError("channel is already closed", [], pos);
  }

  static invalidTokenExpression(pos: Position): SystemError {
    return SystemError.evaluationError("token operator should only be used during parsing", [], pos);
  }

  static invalidApplicationExpression(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_APPLICATION_EXPRESSION, "application on a non-function").withPrimaryLabel(
      "this expression is not a function",
      pos
    );
  }

  static importFailed(name: string, resolved: string, error: unknown): SystemError {
    const err = new SystemError(ErrorType.IMPORT_FAILED, "import failed")
      .withNote(`name: "${name}"`)
      .withNote(`resolved name: "${resolved}"`);
    return error instanceof Error ? err.withNote(`error: "${error.message}"`) : err.withNote(`error: "${error}"`);
  }

  static undeclaredName(name: string, pos: Position): SystemError {
    return new SystemError(ErrorType.UNDECLARED_NAME, `undeclared name ${name}`)
      .withPrimaryLabel("this name is not declared in scope", pos)
      .withNote(`Variable can be declared with ":=" operator like this: ${name} := value`);
  }

  static invalidObjectPattern(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_OBJECT_PATTERN, "object pattern on non-object").withPrimaryLabel(
      "here",
      pos
    );
  }

  static invalidIncrement(name: string, pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_INCREMENT_ASSIGN, "can't increment a non-number variable")
      .withPrimaryLabel(`value of "${name}" is not a number`, pos)
      .withNote(`To use += operator, all names in its pattern should have number values`);
  }

  static invalidIncrementValue(pos: Position): SystemError {
    return new SystemError(ErrorType.INVALID_INCREMENT_ASSIGN, "can't increment by a non-number")
      .withPrimaryLabel("value is not a number", pos)
      .withNote(`To use += operator, value to be incremented by should be a number`);
  }

  static unresolvedImport(name: string, error: unknown): SystemError {
    const err = new SystemError(ErrorType.IMPORT_RESOLVE_FAILED, "can't resolve import").withNote(`name: "${name}"`);
    return error instanceof Error ? err.withNote(`error: "${error.message}"`) : err.withNote(`error: "${error}"`);
  }

  static unbalancedOpenToken(tokens: [string, string], openPos: Position, closePos: Position): SystemError {
    return new SystemError(ErrorType.MISSING_TOKEN, "Missing closing token")
      .withPrimaryLabel(`must be closed with "${tokens[1]}" by that point`, closePos)
      .withPrimaryLabel(`opening token "${tokens[0]}" here`, openPos);
  }

  static unbalancedCloseToken(tokens: [string, string], pos: Position): SystemError {
    return new SystemError(ErrorType.MISSING_TOKEN, `Unexpected closing token "${tokens[1]}"`).withPrimaryLabel(
      "unexpected closing token",
      pos
    );
  }

  static duplicateDefaultExport(pos: Position): SystemError {
    return new SystemError(
      ErrorType.DUPLICATE_DEFAULT_EXPORT,
      "Cannot have multiple default exports in a module"
    ).withPrimaryLabel("unexpected default export", pos);
  }

  static invalidDefaultPattern(pos: Position): SystemError {
    return new SystemError(
      ErrorType.INVALID_DEFAULT_PATTERN,
      "default value for pattern can't be an operator"
    ).withPrimaryLabel("unexpected default pattern", pos);
  }

  static invalidPinPattern(pos: Position): SystemError {
    return new SystemError(
      ErrorType.INVALID_PIN_PATTERN,
      "pin value for pattern can't be an operator"
    ).withPrimaryLabel("unexpected pin pattern", pos);
  }

  static immutableVariableAssignment(patternKey: string, pos: Position): SystemError {
    return new SystemError(ErrorType.IMMUTABLE_VARIABLE_ASSIGNMENT, "expected mutable name").withPrimaryLabel(
      `variable "${patternKey}" is not declared as mutable`,
      pos
    );
  }

  static infixOperatorInPrefixPosition(src: string, op: NodeType, pos: Position): SystemError {
    const precedence = getExprPrecedence(op);
    const isInfix = precedence[1] !== null;
    return new SystemError(ErrorType.MISPLACED_OPERATOR, "infix/postfix operator in prefix position")
      .withPrimaryLabel(`Operator "${src}" can't be used in prefix position`, pos)
      .withNote(
        isInfix
          ? `The "${src}" operator is an infix operator. Provide another operand to the left from it`
          : `The "${src}" operator is a postfix operator. Provide an operand to the left from it`
      );
  }

  static prefixOperatorInRhsPosition(src: string, op: NodeType, pos: Position): SystemError {
    return new SystemError(ErrorType.MISPLACED_OPERATOR, "prefix operator in infix/postfix position").withPrimaryLabel(
      `Operator "${src}" can't be used in infix/postfix position`,
      pos
    );
  }

  static missingOperand(pos: Position): SystemError {
    return new SystemError(ErrorType.MISSING_OPERAND, "missing operand").withPrimaryLabel("here", pos);
  }

  static testError(type: ErrorType): SystemError {
    return new SystemError(type, "test error");
  }
}
