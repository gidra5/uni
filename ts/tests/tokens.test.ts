import { parseMultilineStringToken, parseStringToken, parseToken, specialStringChars } from "../src/parser/tokens.js";
import { beforeEach, describe, expect } from "vitest";
import { it, fc, test } from "@fast-check/vitest";
import { array, type Arbitrary } from "fast-check";
import { Iterator } from "iterator-js";
import { eventLoopYield, getPos } from "../src/utils/index.js";
import { SystemError } from "../src/error.js";
import { position } from "../src/position.js";
import { Injectable, register } from "../src/injector.js";

beforeEach(() => {
  register(Injectable.NextId, 0);
  register(Injectable.PositionMap, new Map());
});

const anyStringArb = fc.string({ size: "large", unit: "binary" });
const commentArb = anyStringArb.filter((s) => !s.includes("\n"));
const blockCommentArb = anyStringArb.map((s) => s.replace("*/", "").replace("/*", ""));
const stringInsidesArb = anyStringArb
  .filter((s) => !s.includes("\n") && !s.includes("\\"))
  .map((s) => s.replace('"', ""));
const charArb = fc.string({ minLength: 1, maxLength: 1 });
const notStringSpecialCharArb = charArb.filter((s) => !specialStringChars.includes(s) && s !== "(");
const arrayLenArb = <T>(arb: Arbitrary<T>, len: number) => fc.array(arb, { minLength: len, maxLength: len });

function dropId({ id, ...token }: any) {
  if ("token" in token) return { ...token, token: dropId(token.token) };
  return token;
}

describe("string token", () => {
  test.prop([stringInsidesArb])("simple string", async (value) => {
    await eventLoopYield();

    const src = `${value}"`;
    const startIndex = 0;
    const expectedToken = { type: "lastSegment", value };
    const expectedIndex = value.length ;

    const [{ index }, token] = parseStringToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  test.prop([
    fc
      .array(stringInsidesArb)
      .chain((strings) => arrayLenArb(notStringSpecialCharArb, strings.length).map((joins) => [strings, joins]))
      .map(([strings, joins]) => {
        const literal = Iterator.zip(strings, [...joins.map((x) => "\\" + x), ""])
          .flat()
          .join("");
        const value = Iterator.zip(strings, [...joins, ""])
          .flat()
          .join("");
        return [literal, value];
      }),
  ])("string token escapes", async ([literal, value]) => {
    await eventLoopYield();

    const src = `${literal}"`;
    const startIndex = 0;
    const expectedToken = { type: "lastSegment", value };
    const expectedIndex = literal.length;

    const [{ index }, token] = parseStringToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect((token as any).value).toEqual(expectedToken.value);
    expect(dropId(token)).toEqual(expectedToken);
  });

  test.prop([
    fc
      .array(stringInsidesArb)
      .chain((strings) => arrayLenArb(notStringSpecialCharArb, strings.length).map((joins) => [strings, joins]))
      .map(([strings, joins]) =>
        Iterator.zip(strings, [...joins.map((x) => "\\" + x), ""])
          .flat()
          .join("")
      ),
  ])("unclosed string token", async (literal) => {
    await eventLoopYield();

    const src = literal;
    const startIndex = 0;
    const expectedIndex = literal.length;
    const expectedToken = {
      type: "error",
      cause: SystemError.unterminatedString(position(startIndex, expectedIndex)),
    };

    const [{ index }, { value: _, ...token }] = parseStringToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  test.prop([
    fc
      .array(stringInsidesArb)
      .chain((strings) => arrayLenArb(notStringSpecialCharArb, strings.length).map((joins) => [strings, joins]))
      .map(([strings, joins]) =>
        Iterator.zip(strings, [...joins.map((x) => "\\" + x), ""])
          .flat()
          .join("")
      ),
  ])("unclosed string token escape", async (literal) => {
    await eventLoopYield();

    const src = `${literal}\\`;
    const startIndex = 0;
    const expectedIndex = literal.length + 1;
    const expectedToken = {
      type: "error",
      cause: SystemError.unterminatedString(position(startIndex, expectedIndex)),
    };

    const [{ index }, { value: _, ...token }] = parseStringToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  test.prop([stringInsidesArb])("unclosed string token before newline", async (literal) => {
    await eventLoopYield();

    const src = `${literal}\n`;
    const startIndex = 0;
    const expectedIndex = literal.length;
    const pos = position(startIndex, expectedIndex);
    const expectedToken = {
      type: "error",
      value: literal,
      cause: SystemError.unterminatedString(pos),
    };

    const [{ index }, token] = parseStringToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  test.prop([
    fc.array(stringInsidesArb).chain((segments) =>
      fc
        .stringMatching(/^\s*$/)
        .filter((s) => !s.includes("\n"))
        .map((s) => "\n" + s)
        .map((intend) => [intend, segments] as const)
    ),
  ])("multiline strings", async ([intend, segments]) => {
    await eventLoopYield();

    const literal = segments.join(intend);
    const value = segments.join("\n").trimEnd();
    const src = `${literal}"""`;
    const startIndex = 0;
    const expectedIndex = literal.length;
    const expectedToken = { type: "lastSegment", value };

    const [{ index }, token] = parseMultilineStringToken(intend).parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });
});

describe("number token", () => {
  it.prop([fc.stringMatching(/^\d+\.\d+$/)])("float literals", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^\d+$/)])("int literals", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^\d+\.$/)])("trailing dot literals", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src + "0") };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect((token as any).value).toEqual(expectedToken.value);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^\.\d+$/)])("prefix dot literals", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number("0" + src) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^\d[\d_]*\d$/)])("int literals with spacers", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src.replace(/_/g, "")) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^\.\d[\d_]*\d$/)])("float literals with spacers", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src.replace(/_/g, "")) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^0x[\da-fA-F][\da-fA-F_]*[\da-fA-F]$/)])("hex literals with spacers", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src.replace(/_/g, "")) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect((token as any).value).toEqual(expectedToken.value);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^0o[0-7][0-7_]*[0-7]$/)])("octal literals with spacers", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src.replace(/_/g, "")) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect((token as any).value).toEqual(expectedToken.value);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^0b[01][01_]*[01]$/)])("binary literals with spacers", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "number", value: Number(src.replace(/_/g, "")) };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect((token as any).value).toEqual(expectedToken.value);
    expect(dropId(token)).toEqual(expectedToken);
  });
});

describe("identifier token", () => {
  it.prop([fc.stringMatching(/^[a-zA-Z]\w*$/)])("regular idents", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "identifier", name: src };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([fc.stringMatching(/^_+$/)])("placeholders", (src) => {
    const startIndex = 0;
    const expectedToken = { type: "placeholder" };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(src, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });
});

describe("comments", () => {
  it.prop([fc.stringMatching(/[^\n]*/)])("single line comments", (comment) => {
    const src = `//${comment}\n`;
    const input = `${src}12412434`;
    const startIndex = 0;
    const expectedToken = { type: "newline" };
    const expectedIndex = src.length;

    const [{ index }, token] = parseToken.parse(input, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([blockCommentArb])("multi line comments", (comment) => {
    const src = `/*${comment}*/`;
    const input = `${src}_`;
    const startIndex = 0;
    const expectedToken = { type: "placeholder" };
    const expectedPos = { start: src.length, end: src.length + 1 };
    const expectedIndex = expectedPos.end;

    const [{ index }, token] = parseToken.parse(input, { index: startIndex });
    const pos = getPos(token.id);

    expect(pos).toEqual(expectedPos);
    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([
    array(
      fc.oneof(
        fc.constant("\n"),
        commentArb.map((comment) => `//${comment}\n`),
        blockCommentArb.filter((t) => !t.includes("\n")).map((comment) => `/*${comment}*/`)
      )
    ).map((parts) => parts.join("")),
  ])("multiple comments", (src) => {
    const input = `${src}_`;
    const startIndex = 0;

    let expectedIndex: number;
    let expectedToken: { type: string };
    if (src.includes("\n")) {
      expectedToken = { type: "newline" };
      expectedIndex = src.length;
    } else {
      expectedToken = { type: "placeholder" };
      expectedIndex = src.length + 1;
    }

    const [{ index }, token] = parseToken.parse(input, { index: startIndex });

    expect(index).toBe(expectedIndex);
    expect(dropId(token)).toEqual(expectedToken);
  });

  it.prop([anyStringArb])("adding line comments instead of newlines never changes the resulting token", (src) => {
    const [, token] = parseToken.parse(src, { index: 0 });
    const [, withComments] = parseToken.parse(withLineComments(), { index: 0 });

    expect(dropId(token)).toStrictEqual(dropId(withComments));

    function withLineComments(): string {
      if (token.type === "newline") {
        const end = getPos(token.id)!.end;
        const tail = src.substring(end);
        return `//\n${tail}`;
      }
      return src;
    }
  });

  it.prop([anyStringArb])("adding block comments never changes the result", (src) => {
    const [, token] = parseToken.parse(src, { index: 0 });
    const [, withComments] = parseToken.parse(`/**/${src}/**/`, { index: 0 });

    expect(dropId(token)).toStrictEqual(dropId(withComments));
  });
});

it.prop([anyStringArb])("parseToken never throws", (src) => {
  expect(() => parseToken.parse(src, { index: 0 })).not.toThrow();
});

it.prop([anyStringArb])("parseStringToken never throws", (src) => {
  expect(() => parseStringToken.parse(src, { index: 0 })).not.toThrow();
});

it.prop([anyStringArb, fc.stringMatching(/\n\s*/)])("parseMultilineStringToken never throws", (src, intend) => {
  const parser = parseMultilineStringToken(intend);
  expect(() => parser.parse(src, { index: 0 })).not.toThrow();
});
