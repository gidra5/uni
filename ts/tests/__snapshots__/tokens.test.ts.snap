// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseTokens 1`] = `
[
  {
    "end": 2,
    "src": "42",
    "start": 0,
    "type": "number",
    "value": 42,
  },
  {
    "end": 10,
    "src": ""Hello"",
    "start": 3,
    "type": "string",
    "value": "Hello",
  },
  {
    "end": 19,
    "src": "variable",
    "start": 11,
    "type": "identifier",
  },
  {
    "end": 21,
    "src": "(",
    "start": 20,
    "type": "identifier",
  },
  {
    "end": 22,
    "src": "(",
    "start": 21,
    "type": "identifier",
  },
  {
    "end": 26,
    "src": "expr",
    "start": 22,
    "type": "identifier",
  },
  {
    "end": 27,
    "src": ")",
    "start": 26,
    "type": "identifier",
  },
  {
    "end": 28,
    "src": ")",
    "start": 27,
    "type": "identifier",
  },
]
`;
